### PR TITLE
BufferAttribute: Support `position` attributes with `itemSize` 2.

### PR DESCRIPTION
**Pull request by Oletus**
**Description**

This enables more tightly packed 2D position data.

*This contribution is funded by [Higharc](https://higharc.com)*

### CODE REVIEW

**Comment by Mugen87** [comment]
We have discussed this topic in the past and afaik the current conclusion is to not add support for two dimensional vertex data. There is a fundamental assumption in the engine that `position` is `vec3`. If we change that, we have to update a lot of library code, see https://github.com/mrdoob/three.js/issues/6288#issuecomment-217512424.

**Comment by Oletus** [comment]
I'm not aware of other parts of the code which would have problems with 2D vertex data. All WebGL shaders should work the same as with 3D data at least. This change was sufficient to support it in our use case, and saving 1/3 of vertex memory can be significant. Any examples of other things that would need to be updated?

**Comment by Oletus** [comment]
True, looking into it a bit more, I think `BufferGeometry.computeTangents()` and `BufferGeometry.applyMatrix4()` are still broken with this version of the patch. I'm pretty sure `computeVertexNormals()` and `computeBoundingSphere()` already work. They should be quite easy to fix and unit test, but if you think this is not worth the maintenance effort, we can close the PR.

**Comment by Mugen87** [comment]
Let's hear the opinions of the others. I'm not sure how everybody looks at this topic in the meanwhile.

**Comment by mrdoob** [comment]
> This enables more tightly packed 2D position data.

Do you have any numbers? Memory savings, ... ?

**Comment by Oletus** [comment]
In our use case the savings are on the order of a few megabytes, and it's easy to imagine more complex 2D data than what we currently have. Seems like this relatively simple change is worth it. I updated the PR with some more fixes and better tests.

File: src/core/BufferAttribute.js [hunk 1]
```diff
@@ -257,6 +257,12 @@ class BufferAttribute {
 
 	getZ( index ) {
 
+		if ( this.itemSize < 3 ) {
+
+			return 0;
+
+		}
```
Comment (Mugen87) [inline src/core/BufferAttribute.js:264]: It think we should also update `setZ()` in a similar fashion for consistency reasons.

File: src/core/BufferAttribute.js [hunk 2]
```diff
@@ -327,7 +333,12 @@ class BufferAttribute {
 
 		this.array[ index + 0 ] = x;
 		this.array[ index + 1 ] = y;
-		this.array[ index + 2 ] = z;
+
+		if ( this.itemSize > 2 ) {
+
+			this.array[ index + 2 ] = z;
+
+		}
```
Comment (Mugen87) [inline src/core/BufferAttribute.js:341]: It think we should also update `setZ()` in a similar fashion for consistency reasons.

**Comment by Mugen87** [comment]
I've played around with this change and something irritates me. I was under the assumption that if you use a position attribute with an item size of `2`, you need to define `vec2` for the position attribute in the shader as well. Otherwise I would expect a WebGL warning like:

> GL_INVALID_OPERATION: Vertex buffer is not big enough for the draw call

However, the following sample code shows no WebGL warning:

```js
import * as THREE from 'three';

const scene = new THREE.Scene();

const camera = new THREE.PerspectiveCamera( 75, window.innerWidth / window.innerHeight, 0.1, 1000 );
camera.position.z = 3;

const renderer = new THREE.WebGLRenderer( { antialias: true } );
renderer.setPixelRatio( window.devicePixelRatio );
renderer.setSize( window.innerWidth, window.innerHeight );
document.body.appendChild( renderer.domElement );

const geometry = new THREE.BufferGeometry();

const vertices = [
	- 1.0, - 1.0,
	1.0, - 1.0,
	1.0, 1.0,
	- 1.0, 1.0
];

const indices = [
	0, 1, 2,
	0, 2, 3
];

geometry.setAttribute( 'position', new THREE.Float32BufferAttribute( vertices, 2 ) );
geometry.setIndex( indices );

const material = new THREE.MeshBasicMaterial();

const cube = new THREE.Mesh( geometry, material );
scene.add( cube );

renderer.render( scene, camera );
```
Live example: https://jsfiddle.net/dtw9z85x/

Can somebody explain why is this the case? Is the `z` component in the shader automatically set to `0` in this case? Is this anywhere defined in the WebGL spec?

**Comment by Mugen87** [comment]
I've implemented the above example with `WebGPURenderer` and the result is the same. The WebGPU spec says:

> Mismatches in the number of components between the vertex format and shader type are allowed, with components being either dropped or filled with default values to compensate. 

I suppose we see the same mechanism in WebGL as well.

Considering this, I think I'm okay with allowing 2D vertex data if a few selected changes to `BufferAttribute` are sufficient. I've verified that bounding sphere and box can be computed with 2D data so I think we should not break any logic that requires bounding volumes.

**Comment by gkjohnson** [comment]
This seems to be explicitly and unambiguously defined behavior in the [OpenGL ES 3.2 specification](https://registry.khronos.org/OpenGL/specs/es/3.2/es_spec_3.2.pdf) upon which WebGL 2 is derived:

> **10.3.3 Transferring Array Elements**
> When a vertex is transferred to the GL by DrawArrays, DrawElements, or the
> other Draw* commands described below, each generic attribute is expanded to four
> components. If size is one then the x component of the attribute is specified by the
> array; the y, z, and w components are implicitly set to 0, 0, and 1, respectively. If
> size is two then the x and y components of the attribute are specified by the array;
> the z and w components are implicitly set to 0 and 1, respectively. If size is three
> then x, y, and z are specified, and w is implicitly set to 1. If size is four then all
> components are specified

File: src/core/BufferAttribute.js (hunk 2 above)
Comment (Mugen87) [reply src/core/BufferAttribute.js:341]: And finally the methods of `InterleavedBufferAttribute` should get the same updates.

**Comment by Mugen87** [comment]
@Oletus I've planned to update your branch but it seems I do not have permission to edit your files. To move forward with this topic, I have created a new PR #27871 and adapted your changes. Of course I'll note you in the release notes if #27871 is accepted and merged.

**Comment by Mugen87** [comment]
Closing in favor of https://github.com/mrdoob/three.js/pull/27871.

**Comment by Oletus** [comment]
@Mugen87 thank you! I didn't have time to update the PR, it's been a very busy week for me. I'm happy as long as the feature lands 🙏

**Pull request closed without merging**